diesel/src/pg/expression/functions.rs

```diff
@@ -2578,3 +2579,101 @@ define_sql_function! {
         Arr: TextArrayOrNullableTextArray + CombinedNullableValue<E,Jsonb>
     >(base: E, path: Arr, new_value: E, create_if_missing: Bool) -> Arr::Out;
 }
+
+#[cfg(feature = "postgres_backend")]
+define_sql_function! {
+    /// Returns target with the item designated by path replaced by new_value,
+    ///     or with new_value added and the item designated by path does not exist.
+    ///
+    /// It can't set path in scalar
+    ///
+    /// All earlier steps in the path must exist, or the target is returned unchanged.
+    /// As with the path oriented operators, negative integers that appear in the path count from the end of JSON arrays.
+    /// If the last path step is an array index that is out of range,
+    ///    the new value is added at the beginning of the array if the index is negative,
+    ///     or at the end of the array if it is positive.
+    ///
+    /// If new_value is not NULL, behaves identically to jsonb_set.
+    ///    Otherwise behaves according to the value of null_value_treatment
+    ///    which must be one of 'raise_exception', 'use_json_null', 'delete_key', or 'return_target'.
+    ///    The default is 'use_json_null'.
```

Comment:
I would rather prefer having an enum with all allowed variants instead of allowing a text input here.

That would require something like this:

https://github.com/diesel-rs/diesel/blob/9861bcba0f48f64481bf7a312da2bb2d3ea2f208/diesel/src/pg/types/mod.rs#L211-L232

and 

https://github.com/diesel-rs/diesel/blob/9861bcba0f48f64481bf7a312da2bb2d3ea2f208/diesel/src/pg/types/ranges.rs#L276-L289

with the corresponding enum variants.

Afterwards the function signature needs to change so that it accepts only the SQL type (`RangeBoundEnum` in that example) instead of arbitrary text. (See here for an example usage: https://github.com/diesel-rs/diesel/blob/9861bcba0f48f64481bf7a312da2bb2d3ea2f208/diesel/src/pg/expression/functions.rs#L442)

Reply:
Thank you for your kind instructions.
I impl to_sql in a new mod called json_function_enum, do you think it's ok?
I'm ready to  make more changes if you think it's not fit to make a new mod for it.

Reply:
No it's good that way, thanks for implementing this change :heart